_tools/types/main.go

```diff
@@ -0,0 +1,333 @@
+	for _, driver := range drivers {
+		if err := analyzeFixtures(driver); err != nil {
+			return err
+		}
+		analyzeCode(driver, uastTypes)
+	}
+
+	formatMarkdownTable(drivers, uastTypes)
+	return nil
+}
+
+type driverStats struct {
+	url          string
+	language     string
+	path         string
+	skip         bool
+	fixturesNum  int
+	fixturesUast map[string]int
+	codeUast     map[string]int
+}
+
+// listDrivers lists all available drivers.
+func listDrivers() ([]*driverStats, error) {
+	fmt.Fprintf(os.Stderr, "discovering all available drivers\n")
+	langs, err := discovery.OfficialDrivers(context.TODO(), &discovery.Options{
+		NoStatic: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+	drivers := make([]*driverStats, 0, len(langs))
+	for _, l := range langs {
+		drivers = append(drivers, &driverStats{
+			language:     l.Language,
+			url:          l.RepositoryURL(),
+			path:         l.RepositoryURL()[strings.LastIndex(l.RepositoryURL(), "/"):],
+			fixturesUast: make(map[string]int),
+			codeUast:     make(map[string]int),
+		})
+	}
+	fmt.Fprintf(os.Stderr, "%d drivers available on-line\n", len(langs))
+	return drivers, nil
+}
+
+// maybeCloneOrPullAll either clones repos to path in local FS or, if already preset,
+// pulls the latest master for each of them.
+func maybeCloneOrPullAll(drivers []*driverStats) error {
+	fmt.Fprintf(os.Stderr, "cloning %d drivers to %s\n", len(drivers), reposRootPath)
+	err := os.MkdirAll(reposRootPath, os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	var (
+		cr = make(chan int, 3)
+		wg sync.WaitGroup
+	)
+	for i := range drivers {
+		wg.Add(1)
+		go maybeCloneOrPullAsync(drivers[i], &wg, cr)
+	}
+	wg.Wait()
+	return nil
+}
+
+func maybeCloneOrPullAsync(d *driverStats, wg *sync.WaitGroup, concurent chan int) error {
+	defer wg.Done()
+
+	concurent <- 1
+	defer func() {
+		<-concurent
+	}()
+
+	return maybeCloneOrPull(d)
+}
+
+func maybeCloneOrPull(d *driverStats) error {
+	repoPath := path.Join(reposRootPath, d.path)
+	_, err := os.Stat(repoPath)
+	if err != nil && !os.IsNotExist(err) {
+		fmt.Fprintf(os.Stderr, "error: %s\n", err)
+		return err
+	}
+
+	if os.IsNotExist(err) {
+		fmt.Fprintf(os.Stderr, "%s does not exist, cloning from %s\n", repoPath, d.url)
+		cmd := exec.Command("git", "clone", d.url+".git")
+		cmd.Dir = reposRootPath
+		err = cmd.Run()
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	fmt.Fprintf(os.Stderr, "%s dir exists, will git pull instead\n", repoPath)
+	cmd := exec.Command("git", "pull", "origin", "master")
+	cmd.Dir = repoPath
+	err = cmd.Run()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type uastType struct {
+	name string
+}
+
+// isUsedIn checks if current UAST type is used in the given package.
+func (u *uastType) isUsedIn() {
+	// TODO
+}
+
+// findUastTypesInSdk finds all types from SDK.
+func findUastTypesInSdk() []uastType {
+	var out []uastType
+	// TODO: load package, iterate all structs
+	types := []interface{}{
```

Comment:
Can be a global.

Reply:
True, but it's intentionally hidden to abstract the way it's being constructed: it's static hard-code for now \w a TODO for building this list dynamically.